Libs/MRML/Core/vtkMRMLScene.cxx

```diff
@@ -4678,3 +4680,46 @@ bool vtkMRMLScene::ParseVersion(const char* versionString, std::string& applicat
   revision = atoi(revisionStr.c_str());
   return true;
 }
+
+//-----------------------------------------------------------------------------
+std::string vtkMRMLScene::ClampFileNameLength(const std::string& filename, int maxFileNameLength/*=25*/, int hashLength/*=4*/)
+{
+  std::string sanitizedBaseName = vtksys::SystemTools::GetFilenameWithoutExtension(filename);
+  if (sanitizedBaseName.length() <= maxFileNameLength)
+  {
+    return filename;
+  }
+
+  // File is too long. Convert it into a shorter filename with the following format:
+  // <first 20 characters of the base name>_<4 character hash code>.<extension>
+  int prefixLength = maxFileNameLength - hashLength - 1; // 1 for the underscore
+  std::string prefix = sanitizedBaseName.substr(0, prefixLength);
+  int suffixLength = sanitizedBaseName.length() - prefixLength;
+  std::string suffix = sanitizedBaseName.substr(prefixLength, suffixLength);
+
+  std::stringstream truncatedFilenameSS;
+  truncatedFilenameSS << prefix << "_";
+
+  // Seed the random number generator with a hash of the suffix
+  std::hash<std::string> hashFunction;
+  size_t hashCodeSeed = hashFunction(suffix);
+
+  vtkNew<vtkMinimalStandardRandomSequence> randomSequence;
+  randomSequence->SetSeed(hashCodeSeed);
+
+  // Generate a random 4 character hash code from 4 upper case letters and numbers
+  std::string hashChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+  for (int i = 0; i < hashLength; ++i)
+  {
+    int index = static_cast<int>(std::floor(std::fmod(randomSequence->GetNextValue(), 1.0) * hashChars.size()));
+    truncatedFilenameSS << hashChars[index];
+  }
+
+  std::string extension = vtksys::SystemTools::GetFilenameExtension(filename);
```

Comment:
Always use the storage node for such operations. vtksys::SystemTools does not know what a file extension is (it only has the option to use the first or last `.`), but cannot recognize actual file extensions.